### DISCUSSION TITLE
Process of converting Next.js examples to TypeScript?

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#33278] [Help]
Reference: https://github.com/vercel/next.js/pull/33245#issuecomment-1012094076

Personally, I'm *not* in favor of converting examples to Typescript. Unless there are some extraordinary benefits to new Next.js projects (which I wager often start as prototytpes), speed of development is the most important metric and Typescript definitely slows this down. 

In addition, Next.js was created with the philosophy of *incredmental adoptability* – including Typescript goes against this philosophy. The framework's name is also `Next.js`, not `Next.ts`. I am in favor of the existing pattern of having Typescript examples alongisde the JavaScript ones (e.g. [blog-starter-typescript](https://github.com/vercel/next.js/tree/canary/examples/blog-starter-typescript)).

What does everyone else think?

**Comment by Dragate** [comment]
I'd prefer if everything was in TS. Speed to development to 1.0.0 might be greater, but it's not fun when client calls for bugs that could have been avoided. Extra time spent beforehand will save tons of time afterwards.

Secondly, I could pretty much understand how to interpret TS examples into vanilla JS code before I learned TS myself. However, there are often times I'm having difficulty how the vanilla JS example would look like in TS. With that being said, I believe TS examples bring more value overall.

**Reply by ghost** [reply]
> often times I'm having difficulty how the vanilla JS example would look like in TS 

Is this not a failure of TS that it requires examples to even understand how to re-implement a vanilla version?

**Reply by AriaFantom** [reply]
Whatever but we before use to just get start with the code cuz all was in js and now if all in Typescript extra time is getting waste for no sucessfull reason. I would high acknowledge to have both in examples or else just have js only, those want ts let them convert, at the name states Next.js Not Next.ts

**Comment by L1lith** [comment]
I find TypeScript to be overly verbose, and as [this article](https://medium.com/javascript-scene/the-typescript-tax-132ff4cb175b) puts it: taxing. Before developing in JavaScript I worked in Java, I since moving I've come to appreciate vanilla JavaScript for being **simple and expressive**. Adding TypeScript into the mix can make it harder to read (and therefore easier for bugs to be hiding). If your code is not legible, adding TypeScript is a band-aid solution. The majority of bugs in my experience are not type-related, but issues related to bad behavior (even if all the types are perfectly correct!). TypeScript does not fix these bugs.

I understand TypeScript may be necessary in **some** use cases, but in my opinion unless you have a specific need for it it's going to increase your development time for very limited rewards. With Next.js being a "batteries-included" framework one of it's specific strengths is how fast development can be. While I understand that means Next.js has opinions, I believe making TypeScript the default moves away from ease of use and towards type safety which may be overkill. 

In Robert C. Martin's book "Clean Code" he describes comments in this way: 

_"Nothing can be quite so helpful as a well-placed comment. Nothing can clutter up a module more than frivolous dogmatic comments. Nothing can be quite so damaging as an old crufty comment that propagates lies and misinformation._

_Comments are not like Schindler’s List. They are not “pure good.” Indeed, comments are, at best, a necessary evil. If our programming languages were expressive enough, or if we had the talent to subtly wield those languages to express our intent, we would not need comments very much—perhaps not at all._


_The proper use of comments is to compensate for our failure to express ourself in
code. Note that I used the word failure. I meant it. Comments are always failures. We must
have them because we cannot always figure out how to express ourselves without them,
but their use is not a cause for celebration."_

I believe that static typing can largely be viewed the same way. To me having to explicit describe each and every type is indicative that our code is simply **not self-explanatory enough**. I can see the benefits of publishing strict typings in a very popular library (or even internally inside of Next.js) for the sake of absolute safety beyond a shadow of a doubt or convenient suggestions the user's IDE, but when we are talking about a small team of developers trying to deploy their website quickly, or just anyone trying to get started with Next.js quickly, having the example code interspersed with (hopefully) redundant typings means more time wasted trying to comprehend it. Having examples preloaded with TypeScript means we're pushing this type system onto the developer.

Furthermore, the time spent reading through and maintaining type definitions could be better spent on more-effective safety measures like TDD, or simply more time spent developing the application. Though I know this isn't a 1-1 comparison, if I was writing a book I wouldn't want to waste my time _verb_ prefixing _determiner_ my _noun_ words.

Edit: Another option would be to separate the JS code and the TypeScript definition themselves

**Reply by ghost** [reply]
Could not have put it better, I completely agree 🙏 especially concerning the library part. Obviously using Typescript in developing Next.js the framework has lot of benefits, but to put that tax on users of the framework is heavy-handed.

**Comment by danielfoxp2** [comment]
> Reference: [#33245 (comment)](https://github.com/vercel/next.js/pull/33245#issuecomment-1012094076)
> 
> Personally, I'm _not_ in favor of converting examples to Typescript. Unless there are some extraordinary benefits to new Next.js projects (which I wager often start as prototytpes), speed of development is the most important metric and Typescript definitely slows this down.
> 
> In addition, Next.js was created with the philosophy of _incredmental adoptability_ – including Typescript goes against this philosophy. The framework's name is also `Next.js`, not `Next.ts`. I am in favor of the existing pattern of having Typescript examples alongisde the JavaScript ones (e.g. [blog-starter-typescript](https://github.com/vercel/next.js/tree/canary/examples/blog-starter-typescript)).
> 
> What does everyone else think?

I fully agree with you. I was following a tutorial that used the example "with-tailwindcss" and created an app with normal, vanilla js. Then, for my surprise, the one I got was fully configured with typescript. I don't like typescript and rather do my things in js itself. So I'm losing a big time trying to figure it out how to get back to the previous behavior and create my app with vanilla js.

The examples shouldn't be so opinionated. It should be an user choice. If one want something in ts, then it should be a ts example for it.
It is kind of dictatorial move to force people something. Freedom is a matter of individual choice, not someone else deciding for you what they think is better for you.

Can someone help me how to create an app with the previous behavior? If I fork the previous way and point the create-next-app to the forked repo would it work? I'm completely new on this. Thanks everyone.

p.s.: This doesn't mean the maintainers shouldn't enforce ts usage. They should if they want. But it shouldn't change well known behaviors like that. 

p.s.2: By the way, for tailwindcss, there is an official way to install it together with Next.js in the link below, if somebody needs it: 
https://tailwindcss.com/docs/guides/nextjs

**Comment by Dueen** [comment]
I think adding examples in Typescript aligns 100% with the Next.js philosophy. 

Next.js calls itself 'The React Framework for Production', If it were to focus it's efforts on making developers feel comfortable with getting to boilerplate code than it should drop the production part of that tagline. 
But that's not the goal, the goal is to make your app (code) production ready. Typescript makes this easier, although admittedly with some initial hurdles to jump over. But these are hurdles you will have to deal with either before you ship or after. I prefer to do so before.

**Reply by danielfoxp2** [reply]
> I think adding examples in Typescript aligns 100% with the Next.js philosophy.
> 
> Next.js calls itself 'The React Framework for Production', If it were to focus it's efforts on making developers feel comfortable with getting to boilerplate code than it should drop the production part of that tagline. But that's not the goal, the goal is to make your app (code) production ready. Typescript makes this easier, although admittedly with some initial hurdles to jump over. But these are hurdles you will have to deal with either before you ship or after. I prefer to do so before.

Are you implying that vanilla js it is not good or ready for production? I can't agree with your statement. In my case, the ts made it harder for me to be ready for production. It is slowing me down a lot, right now. Vanilla js is also 100% aligned with Next.js philosophy. 

And it is all fine that you have your preferences and you should have a "with-tailwindcss-ts" to help you achieve your goals the way you like. But who prefer to use vanilla js, the well known "with-tailwindcss" should still work as expected. Typescript is not for everyone and it shouldn't be assumed as so.

**Reply by Dueen** [reply]
Of course vanilla JS is ready for production. Typescript compiles to vanilla JS at the end of the day. 
My argument is on the benefits vs the pitfalls of Typescript. If you're shipping production code, having that extra benefit of type safety at compile time is a net benefit to the extra coding time you have to do to make Typescript happy. So ultimately you can focus on the complicated bugs that Typescript can't catch, and less on the ones that Typescript could've gotten for you.
